File: chapter_1_arrays_strings/js/1.4-palindrome.js
const isAlpha = ch => {
  // case insensitive regex
  return /^[A-Z]$/i.test(ch)
}

// O(n) solution
const isPaliPermutation = str => {
  const obj = {}
  let odds = 0
  for (let ch of str) {
    if (isAlpha(ch)) {
      obj[ch] == undefined ? (obj[ch] = 1) : obj[ch]++
      obj[ch] % 2 != 0 ? odds++ : odds--
    }
  }
  return odds <= 1
}

console.log(isPaliPermutation("tact coa")) // true
console.log(isPaliPermutation("tact coao")) // true
console.log(isPaliPermutation("tact ccoa")) // false
console.log(isPaliPermutation("tact coaz")) // false
console.log(isPaliPermutation("at ta")) // true
console.log(isPaliPermutation("t")) // true

File: chapter_1_arrays_strings/cpp/1.4-palindrome.cpp
#include <iostream>
#include <string>
#include <ctype.h>

using namespace std;

// O(n) solution
bool is_pali_permutation(string str)
{
    const int ASCII_LENGTH = 128;
    int map[ASCII_LENGTH] = {0};
    int odds = 0;

    for (auto ch : str)
    {
        if (isalpha(ch))
        {
            map[ch]++;
            map[ch] % 2 != 0
                ? odds++
                : odds--;
        }
    }
    return odds <= 1;
}

int main()
{
    cout << is_pali_permutation("tact coa") << endl;  // 1
    cout << is_pali_permutation("tact coao") << endl; // 1
    cout << is_pali_permutation("tact ccoa") << endl; // 0
    cout << is_pali_permutation("tact coaz") << endl; // 0
    cout << is_pali_permutation("at ta") << endl;     // 1
    cout << is_pali_permutation("t") << endl;         // 1
    return 0;
}